package.json

```diff
@@ -39,23 +41,16 @@
     "build:plugin:webp-uploads": "webpack --mode production --env plugin=webp-uploads",
     "format-js": "wp-scripts format",
     "lint-js": "wp-scripts lint-js",
-    "format-php": "composer format",
     "phpstan": "composer phpstan",
-    "lint-php": "composer lint:all",
+    "lint:php": "composer lint",
+    "lint:php:fix": "composer format",
+    "lint:php:plugins": "cross-env WPP_PLUGIN=$npm_config_plugin php -f bin/phpcs/lint.php",
+    "lint:php:plugins:fix": "cross-env WPP_FIX=1 WPP_PLUGIN=$npm_config_plugin php -f bin/phpcs/lint.php",
```

Comment:
Why do we need these here? What prevents us from implementing this as a PHP-based command in `composer.json`?

A more specific question, so that I understand correctly: Where does the `$npm_config_plugin` variable come from here?

Reply:
The `$npm_config_plugin` variable comes from the `--plugin` command line arg to `npm`. https://dev.to/felipperegazio/handling-command-line-arguments-in-npm-scripts-2ean

Reply:
> Why do we need these here? What prevents us from implementing this as a PHP-based command in composer.json?

@felixarntz We can do it, but then if we run `lint:plugins` from npm and it requires to pass PHPCS args, we will need to escape the args twice e.g. `npm run lint:php:plugins -- -- --help`.

Reply:
@felixarntz @westonruter Can you please share your opinion here?

Reply:
If we can avoid the extra `--` in the commands, I'm all for it!

Reply:
> > Why do we need these here? What prevents us from implementing this as a PHP-based command in composer.json?
> 
> @felixarntz We can do it, but then if we run `lint:plugins` from npm and it requires to pass PHPCS args, we will need to escape the args twice e.g. `npm run lint:php:plugins -- -- --help`.

IMO we shouldn't have to use NPM to run those PHP scripts in the first place. If we implement this with Composer, we can probably find a way to do it like `composer lint:plugins -- --plugin=webp-uploads`.

Of course we could also consider making it work well for both. Wouldn't we be able to use `npm run lint:php:plugins --plugin=webp-uploads` by having that command call `composer lint:plugins -- --plugin=$npm_config_plugin`? With something along those lines, we could get the best of both worlds.

Reply:
> IMO we shouldn't have to use NPM to run those PHP scripts in the first place. If we implement this with Composer, we can probably find a way to do it like `composer lint:plugins -- --plugin=webp-uploads`.

Isn't it better for DX to have all the commands available in either `npm` _or_ `composer` but not both? It is not helpful to have to try to remember whether a task is accomplished with `npm` or `composer` each time you need to run it. If we do have the scripts in `composer.json`, then I think they should just be there for redundancy. But with redundancy, then there is more maintenance overhead. 

Also, if we can avoid having to add that extra `--` it will also be helpful. I don't know about you, but I often forget and then it doesn't work as expected, so then I have to go back and remember to add it, and put it in the right spot.

Reply:
@felixarntz As you mentioned in https://github.com/WordPress/performance/pull/1090#discussion_r1548554987, what if we keep NPM as a wrapper inside the composer? It will work well in that case IMO. I understand your apprehension about including PHP-specific elements in Composer. However, in terms of practical execution, the process remains the same whether we call it within NPM or Composer. It's essentially an implementation detail that shouldn't disrupt any workflows, as the `lint:php` command will perform its intended function regardless of the script runner in which it's executed.

 Inside composer scripts, add `"lint:plugins": "npm run lint:php:plugins"` and we can call it like:
```
composer lint:plugins # Lint all plugins.
composer lint:plugins -- --plugin=foo # Lint specific plugin
composer lint:plugins -- --plugin=auto-sizes -- --help # Pass args to PHPCS.
```

Reply:
> Isn't it better for DX to have all the commands available in either `npm` _or_ `composer` but not both? It is not helpful to have to try to remember whether a task is accomplished with `npm` or `composer` each time you need to run it.

I'm not saying we have to have them in both, that was just to try to satisfy both your and my concerns. To me, there's a simple way to separate without having to remember one by one: Commands for PHP go into `composer.json`, while commands for JS or other (not PHP) environment related things go into `package.json`.

Reply:
Again, the bigger question and what we need to further discuss in #1012 is whether the approach of this PR makes sense. Trying to solve both developing from the root directory and developing from the individual plugin directory already shows as of this PR how many workarounds it requires and headaches it gives due to those workarounds being controversial to some contributors.

While my preference for various reasons is to encourage development from the root directory, either way that we decide for the solution may differ:
* If everyone contributes to this from the root directory, then I'm not sure whether this change is needed.
* If everyone contributes from the individual plugin's directory, then we should simply have `composer.json` files per plugin directory (similar to how other monorepos of that approach do it) which would make it work without such workarounds.